core/executor/src/allocator.rs

```diff
@@ -196,7 +245,21 @@ mod tests {
 		let ptr = heap.allocate(1).unwrap();
 
 		// then
-		assert_eq!(ptr, 8);
+		// returned pointer must start right after `PREFIX_SIZE`
+		assert_eq!(ptr, PREFIX_SIZE);
+	}
+
+	#[test]
+	fn should_use_minimum_allocation_size_for_zero_allocations() {
+		// given
+		let mem = MemoryInstance::alloc(Pages(1), None).unwrap();
+		let mut heap = FreeingBumpHeapAllocator::new(mem, 0);
+
+		// when
+		let ptr = heap.allocate(0).unwrap();
```

Comment:
Why do we return a valid pointer for a zero size allocation?

I think we should return a null pointer.

https://stackoverflow.com/questions/1073157/zero-size-malloc

https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#safety-1

Or some sentinel address, but the caller is already doing something incorrect, when he wants to allocate `0` bytes.

Reply:
Zero-sized `malloc`ations are legal and they can return any value they wish. The only requirement is that this value should be accepted by `free`. The stackoverflow answer mentions that.

As to `GlobalAlloc`, returning 0 from `alloc` is treated as an out-of-memory error by its callers. So typically the implementations of `GlobalAlloc` that forward calls to some sort of `malloc` they would have to explicitly check the result of malloc for null or for the size of allocation.

In our case, we don't do it, because we rely on the property that `ext_malloc` returns non zero addresses.

>  but the caller is already doing something incorrect, when he wants to allocate 0 bytes.

In my experience I have never seen calls to `alloc` with zero size in Rust. However, I've seen some "weird" code, like `memcpy` copying something from invalid address, which should be generally UB, however, in that case the length of the copy was zero ¯\_(ツ)_/¯ . So I wouldn't be surprised if I see zero sized allocations in Rust.

Reply:
I think this warrants some docs actually.

Reply:
The rust documentation itself says:
> This function is unsafe because undefined behavior can result if the caller does not ensure that layout has non-zero size.

So, I would not expect any valid rust code to allocate zero bytes.

And further the rust documentation says:
> Returning a null pointer indicates that either memory is exhausted or layout does not meet allocator's size or alignment constraints.

Whatever we do, I would not recommend to actually allocate 8 bytes and return this memory space. We should go for returning some sentinel address, like the address of our base address or something similar. This could be checked in `free` as well.

Reply:
Ah, good point regarding non-zero layout requirement. But yeah, I'd agree that actual allocation doesn't make a lot of sense. I am not sure if we should fix it within this PR though.

Reply:
I removed the test again and created a follow-up ticket where further discussion can take place: https://github.com/paritytech/substrate/issues/3561.